### ISSUE TITLE
CGROM Set ROM B?

### ISSUE THREAD

**Issue by StevenM1** [jafrado/2004_i2c_oled#1]
Hi,

This isn't really an issue, but more of a question that I hope you can answer: I was wondering if you know how to select CGROM Set B in this display? I got the exact same display which works fine, but the characters in ROM B would suit my use better. I kinda expected to set it in `lcd_2004_nuttx.c` lines 588-595:

https://github.com/jafrado/2004_i2c_oled/blob/a6219c32d7a2fb84b330a4a60e1feb5d8a21aada/lcd_2004_nuttx.c#L588-L595

by setting 0x01 instead of 0x00 in line 595. That's mostly based on the code I found here in this other repository:

https://github.com/n2j7/SSD1311/blob/f830a376dfce0b00d0853a4db7d0c04efd6689c8/SSD1311.cpp#L249-L270

But first setting 0xDC followed by 0x01 doesn't seem to do the trick. I also noticed that the other repository sets 0x72 prior to sending the data (and that's also mentioned in your code, but after an `if 0`), but setting 0x72 followed by 0x01 also doesn't seem to work...

Any clue how to do this? I'll admit that I only half know what I'm doing here so I may be missing something obvious..
Thanks!

**Comment by jafrado** [comment]
Goedemorgen, and greetings from sunny California!

Right, the comment is wrong and I need to clean this up, I have a more recent version of the init code for the EA OLED displays, I have just not had free time to regression test and I want to spin the board but life, work and other things have taken priority :)

The definitive reference for the Wide.HK 20x4 OLED display is the SSD1311 manual, you can find it in the `doc` directory of your checkout or here [SSD13311](https://github.com/jafrado/2004_i2c_oled/blob/master/doc/SSD1311_IC_SPEC.pdf). The code is meaningless jibberish for anyone without that in hand.

The SSD1311 has 3 sets of CGROM (ROM A / B / C) software or hardware pin selectable. So the first action for ROM-B is to determine that the board is not hard-wired for the CGROM-A, no matter of programming will fix a HW strap. According to page 4, the `OPR[1:0]` pins determine which CGROM is in effect. So first, you must probe these pins and check it is not hardwired before any programming exercises. That is an artifact of the confusion you see around the net (my board is HW strapped for ROM A), so that is why it is commented out.

The programming is a little bit weird. generally - command, then some number of bytes of data. This is virtualized over I2C which makes the programming even more confusing as it is indirect. Lucky for you, I coded `ssd13xx_command` so, for example to clear the screen, one uses `lcd_cls` which call` ssd_13xx_command `with bit 1 set.

For the CGROM, Table 5-2 on Page 22, we send the first line which is command `0x72`, followed by the data bits. So we would write (``D3=ROI1=0,D2=ROI0=0,D1=0,D0=1``) which is 0x1 for ROM-A.  For ROM B, RO[1:0]=01b, so we would write (`D3=ROI1=0,D2=1,D1=0,D0=1`) which is 0x5. So you should try to send:

```
ssd13xx_command(0x72); /* **** Set ROM */
ssd13xx_command(0x05); /*  **** Set ROM B and 8 CGRAM */


```

There is an important note on that page of the datasheet:

```
"Note: It is recommended to turn off the disply (cmd 08h) before setting no. of CGRAM and defining character ROM, while clear display (cmd 01h) is recommended to sent afterwards."

```
This is whence it should be done, e.g. order of operations, you either program this inline as we bring-up and turn on the display or you should turn off the display, select CGROM, and then clear it to prevent issues (e.g. HW lock up and/or junk on the display from the scan converter). 

Table 4-5 describes how to make the CGRAM characters; you may find it more useful to use this instead to define your own characters versus switching the CGROM. It's only a matter of defining your bit-pattern for the CGRAM and then you can blast out your own characters. To do this, look at `lcd_define_nav_symbols` for how I created some navigation symbology.

Good luck and let me know what you find!

**Comment by StevenM1** [comment]
Good morning, or goedemorgen,

Thank you so much for this very detailed and helpful response. I'm quite far outside of my comfort zone here (it's a fun hobby project - but I do lack some quite fundamental knowledge here and there), and your reply really helped me on the way to understanding what I am actually doing.

I'll soon have a look at my hardware in more detail, but before I start tinkering around, just one quick follow-up question for clarification. I think I now better understand how the instruction sending works; we first send 01110010 = 0x72 so the module knows what to do with the the subsequent data bits, followed by the actual bits that define the to-be-selected ROM. These data bits are either 0001 = 0x01 for ROM-A, or 0101 for ROM-B. But here I'm confused: you wrote that the latter equals 0x03? Why is that not 0x5 (which is 0101 in binary, right)?

As you can probably tell, I really am just catching up on the basics, so please forgive my ignorance here if this is a stupid question :-) 

Thanks again!
Cheers

**Comment by jafrado** [comment]
yes, that is right. Yes, sorry, I typo'd that, it is 0x05, please try the code above. The datasheet does not have the pin-out, I found a more recent copy from EA - https://www.lcd-module.de/fileadmin/eng/pdf/zubehoer/ssd1311.pdf

Look for those pins under a microscope or magnifying glass and find out where they go, when powered, you can probe the pins to see their logic level and determine the strapping. Some boards have a little resistor next to them you can change.

**Comment by StevenM1** [comment]
Hmm, I've been trying to understand where to find the OPR[0:1] pins, but I'm a bit lost to be honest. According to page 10 of the most recent datasheet version, there should be two rows of die pads, and the OPR pins should be located somewhere on the bottom right (if I follow the coordinates on page 11 correctly: 1410,-495 and 1530,-495). However, when I look at my module, I don't see any of those die pads (also tried a magnifying glass). I fear I'm looking at the wrong side (ie am I supposed to remove the glass display from the front side to reveal these pins? The glass is glued on the board as far as I can tell so I'm a bit hesitant to remove it)

![IMG_20210730_103234164 2](https://user-images.githubusercontent.com/17780626/127735657-65868db4-c264-4791-b816-0591e6c5c219.jpg)

![IMG_20210730_103245488 2](https://user-images.githubusercontent.com/17780626/127735663-85f2afd1-9920-48bd-b23b-3ff6c64f9825.jpg)

Reading further through the new datasheet, I also noticed on page 13 and 34 that the CGROM is selected by the ROM0 and ROM1 pins (as opposed to the OPR1 and OPR0 pins which was suggested by the previous version of the data sheet, which instead select the CGRAM). Both should be 1 in order for the ROM to be software selectable. The rear of my module does indeed show ROM0 and ROM1 pins (just left of the SCL/SDA/GRND/VCC) pins, and those appear clean (contrary to this person's display: https://forum.arduino.cc/t/i2c-oled-display-with-5-analog-inputs-on-arduino-uno/349958/11 ; where the ROM0/1 and OP0/1 pins appear soldered; not sure why the exterior-most pins are bridged, I suppose that depends on the wiring on the front side of the board..)

In light of my lack of understanding how to figure out whether there's a hardwired limit to selecting the ROM, I just went ahead and tried programming ROM selection. Firstly I tried your suggestion in the init sequence:

```
ssd13xx_command(0x72); /* **** Set ROM */
ssd13xx_command(0x05); /*  **** Set ROM B and 8 CGRAM */
```

Which didn't work, but then I realised that (I think) 0x05 should (probably) not be sent as a command but as data, so I tried:

```
ssd13xx_command(0x72); /* **** Set ROM */
ssd13xx_data(0x05); /*  **** Set ROM B and 8 CGRAM */
```

but this unfortunately also didn't work. But as far as I can tell from the data sheet, these commands should be correct, so it gives me the impression that also my display is hardwired for ROM A. Which is a bit of a bummer but like you say, I could also define my own characters.

**Comment by jafrado** [comment]
Outstanding! Ok, so you have the same board as I do! I have those pads soldered to select ROM-A. But that is ok, you do not have a jumper selection (you would splash a little solder on either side to select a high/low setting). Right, good catch on the data cycle, that is also needed for this command. But what is more important, which I neglected to tell you, is that the state of the IS, SD, RE bits also needs to be set. Let me check in a fix which will address this and also let you select any internal ROM (A, B, or C).

**Comment by StevenM1** [comment]
Awesome, it works like a charm! Thanks for your detailed replies and help; it not only works, but I also learned a lot from this!

**Comment by jafrado** [comment]
my pleasure - enjoy; do send me some pictures when you get your project working!

**Comment by StevenM1** [comment]
The project is almost done, I'm basically just making an internet radio (and media center) so nothing very spectacular. It uses the libreelec distribution, I set it up such that it automatically plays internet radio on boot. The display is very useful because it allows me to see which channel (and song/associated metadata) is playing, without having to turn on the tv. And OLED just looks so much prettier than standard LCD displays, which would've been the easier option. I additionally installed an IR sensor so I can control it (switch channels, mute, or the media center options) with an old remote I had lying around. The channel/metadata is fed to the display using lcdproc, I hacked around those drivers to make it control this display as well (which worked thanks to your code).

![IMG_20210802_111230993](https://user-images.githubusercontent.com/17780626/127844756-cbc2f61e-5718-41cd-af50-8d6756b5ecc0.jpg)

ROM B was interesting to me because I like to listen to classical music, and a lot of the composers and compositions have accented characters in their names that aren't featured in the character set in ROM A (Antonín Dvořák, Bartók, Leoš Janáček, German composers/pieces with ö ä and the like, some Scandinavians with å). Of course there was always the option to simplify their names or simply get rid of those accents - but what's the fun in that? :) Alternatively, I could indeed create these characters myself in the CGRAM but that would've been trickier to code into the lcdproc drivers..

So, it's now a working internet radio and media center. It's an Rpi4 with, as additional hardware, a heat sink + fan enclosure, an IR sensor, and the display. All I need to do now is make an enclosure, not sure how yet but I'll think of something functional. Quite happy with it since I was still listening to good old analogue radio, but the cable companies are dropping support for that here soon so I had to look for an alternative. I could of course have simply bought a digital radio, but again, what's the fun in that? :)
